feat: added support for index include

lohart13 opened the pull request:
<!--
Thanks for wanting to fix something on Sequelize - we already love you!
Please fill in the template below.
If unsure about something, just do as best as you're able.

If your PR only contains changes to documentation, you may skip the template below.
-->

### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

<!-- NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open. -->

### Description Of Change

This pull requests adds the ability to support indexes with included columns for MSSQL.
For information on how this works and the benefits, please see [Microsoft documentation](https://docs.microsoft.com/en-us/sql/relational-databases/indexes/create-indexes-with-included-columns?view=sql-server-ver15) of this feature.

Fixes #15124

WikiRik commented on src/dialects/mssql/index.js:
```diff
@@ -37,6 +37,7 @@ export class MssqlDialect extends AbstractDialect {
       type: true,
       using: false,
       where: true,
+      include: true,
```
You should add the include option in the abstract index as well;
https://github.com/sequelize/sequelize/blob/fe3f470194e41677c25504910cb5458a70382529/src/dialects/abstract/index.ts#L82-L92
and 
https://github.com/sequelize/sequelize/blob/fe3f470194e41677c25504910cb5458a70382529/src/dialects/abstract/index.ts#L195-L205

lohart13 replied:
Added now

WikiRik reviewed (commented):
Will need to do a more in-depth review later, but these are my initial comments

lohart13 commented on test/unit/sql/index.test.js:
```diff
@@ -253,6 +253,18 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
+          fields: ['email'],
+          include: ['first_name', 'last_name'],
+        }), {
+          mssql: 'CREATE INDEX [email_include_name] ON [User] ([email]) INCLUDE ([first_name], [last_name])',
```
Updated now

lohart13 commented on test/integration/dialects/mssql/errors.test.js:
```diff
@@ -0,0 +1,68 @@
+const { DataTypes, Sequelize } = require('@sequelize/core');
+
+if (dialect === 'mssql') {
+  describe('[MSSQL Specific] Errors', () => {
```
I've changed this to use the `dialect.supports` object

```js
    if (!current.dialect.supports.index.include) {
      return;
    }
```

lohart13 commented:
@WikiRik looking at the test, do you want this to exist under the dialects > mssql folder?
It appears that db2 and postgres can support index with include (in specific versions) therefore not exactly mssql specific.

- https://www.postgresql.org/docs/current/sql-createindex.html
- https://www.ibm.com/docs/en/db2/11.1?topic=statements-create-index

WikiRik commented:
I guess you can include the tests after the following;
https://github.com/sequelize/sequelize/blob/79a53009e86df7efe12642aaffaa36a14bd8abc7/test/integration/model.test.js#L612

But I might prefer a separate test file `test/integration/index.test.js` since model.test.js already has quite some tests

lohart13 commented:
> But I might prefer a separate test file `test/integration/index.test.js` since model.test.js already has quite some tests

I've updated the test location now.
Let me know if you have any further questions

WikiRik commented:
Can you enable the support for db2 and postgres in this PR as well?

lohart13 commented:
@WikiRik I've added support now, however there are some conditions for postgres and db2.

For postgres, the include option for indexes is only supported in version 11 and above therefore this would not be supported for all the versions of postgres that sequelize currently support.

For db2, the include option for indexes is only available for unique indexes.

I've added in some errors and tests to make sure that these limitations are enforced.

Let me know if you have any questions/suggestions

WikiRik commented:
Thanks! I think we're dropping support for postgres 10 in v7 as well since the final release of that will come out on November 10th so let's wait with merging this PR until we do so. That makes the implementation easier.

Pinging @ephys @sdepold as well to confirm dropping postgres 10 support as soon as postgres itself does it

ephys commented:
We can drop postgres 10 in v7 
![image](https://user-images.githubusercontent.com/1280915/200132656-87f26ec2-13ae-4161-b033-542b2d97f63e.png)

WikiRik commented:
I've created a PR to drop the postgres 10 support; it's #15289. Once that has been merged, this PR can be updated accordingly

lohart13 commented:
I've removed the postgres db version checks now.
LMK if there are any further changes required

ephys commented on test/unit/sql/index.test.js:
```diff
@@ -232,6 +233,38 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
       });
     }
+
+    if (current.dialect.supports.index.include) {
```
this test should run on all dialects, but dialects that don't support it should check that an error is thrown. You can do it like this:

```ts
// arrow function so expectsql can catch the error
expectsql(() => sql.addIndexQuery('User',  {
  name: 'email_include_name',
  fields: ['email'],
  include: ['first_name', 'last_name'],
  unique: true,
}), {
  'mysql mariadb': new Error('Your error message here')
  postgres: 'CREATE UNIQUE INDEX [email_include_name] ON [User] ([email]) INCLUDE ([first_name], [last_name])',
});
```

lohart13 replied:
Done

ephys commented on test/unit/sql/index.test.js:
```diff
@@ -232,6 +233,38 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
+        expectsql(sql.addIndexQuery('User',  {
+          name: 'email_include_name',
+          fields: ['email'],
+          include: ['first_name', 'last_name'],
```
could you add a test where `include` is a literal, and `include` is an array of literals? We try to support literal everywhere as an escape hatch

```ts
include: [literal('first_name')],

//

include: literal('(first_name, last_name)'),
```

it should add them as-is without further processing

lohart13 replied:
Done

lohart13 commented on src/dialects/abstract/query-generator.js:
```diff
@@ -667,6 +668,11 @@ export class AbstractQueryGenerator {
     });
 
+    let includeSql;
+    if (this.dialect.supports.index.include && options.include) {
```
Updated now and added support for literals

lohart13 commented on test/integration/index.test.js:
```diff
@@ -0,0 +1,139 @@
+const dialect = Support.getTestDialect();
+const { DataTypes, Sequelize } = require('@sequelize/core');
+
+describe(Support.getTestDialectTeaser('Indexes'), () => {
```
Done

ephys reviewed (dismissed):
This looks like really good work! Thank you :)

WikiRik commented on test/unit/sql/index.test.js:
```diff
@@ -232,6 +233,61 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
+        fields: ['email'],
+        include: ['first_name', 'last_name'],
+      }), {
+        // DB2 does not support non-unique indexes with include columns
```
```suggestion
```

WikiRik commented on test/unit/sql/index.test.js:
```diff
@@ -232,6 +233,61 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
+        fields: ['email'],
+        include: literal('(first_name, last_name)'),
+      }), {
+        // DB2 does not support non-unique indexes with include columns
```
```suggestion
```

WikiRik commented on test/unit/sql/index.test.js:
```diff
@@ -232,6 +233,61 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
+        fields: ['email'],
+        include: [literal('first_name'), literal('last_name')],
+      }), {
+        // DB2 does not support non-unique indexes with include columns
```
```suggestion
```

WikiRik commented on test/integration/index.test.ts:
```diff
@@ -0,0 +1,126 @@
+        indexes: [{ name: 'user_username', fields: ['username'], include: ['first_name', 'last_name'], unique: false }],
+      });
+
+      // DB2 does not support non-unique indexes with include columns
```
```suggestion
```

WikiRik reviewed (approved):
These comments are redundant since the error is clear. But it's a non-blocking issue.
Thanks for working on this!

krzysztofw-hippo commented:
Hi, any idea when this will be released? I don't see it in v6.26.0, will it be included in v6.26.1?
Many thanks

WikiRik commented:
This will land in the next v7 alpha, there are no plans from the maintainers to backport this to v6 but we're always open to PRs from the community

lohart13 commented:
Tagging PR to https://github.com/sequelize/website/issues/501
